pulser-core/pulser/register/_coordinates.py

```diff
@@ -24,41 +28,45 @@ class CoordsCollection:
         _coords: The coordinates.
     """
 
-    _coords: np.ndarray | list
+    _coords: pm.AbstractArray | list
 
     @property
     def dimensionality(self) -> int:
         """The dimensionality of the coordinates (2 or 3)."""
         return self._sorted_coords.shape[1]
 
     @property
-    def sorted_coords(self) -> np.ndarray:
+    def sorted_coords(self) -> pm.AbstractArray:
         """The sorted coordinates."""
         # Copies to prevent direct access to self._sorted_coords
         return self._sorted_coords.copy()
 
+    @cached_property
+    def _rounded_coords(self) -> pm.AbstractArray:
+        coords = pm.vstack(cast(Sequence[AbstractArrayLike], self._coords))
+        return pm.round(coords, decimals=COORD_PRECISION)
+
     @cached_property  # Acts as an attribute in a frozen dataclass
-    def _sorted_coords(self) -> np.ndarray:
-        coords = np.array(self._coords, dtype=float)
-        rounded_coords = np.round(coords, decimals=COORD_PRECISION)
+    def _sorted_coords(self) -> pm.AbstractArray:
         sorting = self._calc_sorting_order()
-        return cast(np.ndarray, rounded_coords[sorting])
+        return self._rounded_coords[sorting]
 
     def _calc_sorting_order(self) -> np.ndarray:
         """Calculates the unique order that sorts the coordinates."""
-        coords = np.array(self._coords, dtype=float)
         # Sorting the coordinates 1st left to right, 2nd bottom to top
-        rounded_coords = np.round(coords, decimals=COORD_PRECISION)
-        dims = rounded_coords.shape[1]
-        sorter = [rounded_coords[:, i] for i in range(dims - 1, -1, -1)]
+        dims = self._rounded_coords.shape[1]
+        sorter = [
+            self._rounded_coords.as_array(detach=True)[:, i]
```

Comment:
This is detaching for every `i`, it would be better to detach only once and store the array in a variable